proposals/0000-class-method-headers.rst

```diff
@@ -0,0 +1,437 @@
+   declaration::
+
+     class C a where
+       f :: a         -- note: no @a
+
+     -- f :: forall a. C a => a
+
+     instance C Int where
+       f @Int = 42    -- still allowed
+
+#. Class variables bound in methods always precede other arguments.
+
+Examples
+--------
+
+* Comma-separated class methods with different class variable ordering::
+
+    class C a b where
+      f @a @b, g @b @a :: a -> b
+
+    -- f :: forall a b. C a b => a -> b
+    -- g :: forall b a. C a b => a -> b
+
+* Erroneous class method header that mentions non-class variable::
+
+    class C a where
+      f @b :: a -> b
+
+  Rejected with the following message::
+
+    • ‘b’ is not bound in the class declaration header ‘C a’
+    • In the class method header: f @b
+
+  The implementation may opt to provide a different error message in the same spirit.
+
+* Erroneous class method header that mentions class variables out of dependency order::
+
+    class C (a :: k) where
+      f @a @k :: P a
+
+    -- f :: forall a k. C (a :: k) => P a
+
+  Rejected with the following message::
+
+    • These kind and type variables: a k
+      are out of dependency order. Perhaps try this ordering:
+        k (a :: k)
+    • In the compiler-generated class method signature:
+        f :: forall a k. C (a :: k) => P a
+
+  This is the same message as one would get if this signature was written by hand.
+  The implementation may opt to provide a different error message in the same spirit.
+
+* Class method header that results in visible forall::
+
+    class C (a :: k) where
+      f @k a :: P a
+
+    -- f :: forall k. forall a -> C (a :: k) => P a
+
+* Instance declaration where the method definition binds class variables::
+
+    class C a b where
+      f :: a -> b
+
+    -- f :: forall a b. C a b => a -> b
+
+    instance C Int Bool where
+      f @Int @Bool = even
+
+* Instance declaration where the method definition binds some, but not all of class variables::
+
+    class C a b where
+      f @a @b :: a -> b
+
+    -- f :: forall a b. C a b => a -> b
+
+    instance C Int Bool where
+      f @Int = even
+
+* Instance declaration where the method definition binds class variables in a different order::
+
+    class C a b where
+      f @b @a :: a -> b
+
+    -- f :: forall b a. C a b => a -> b
+
+    instance C Int Bool where
+      f @Bool @Int = even
+
+* Erroneous instance declaration where the class variable bound in a method definiton does not match::
+
+    class C a where
+      f @a :: a
+
+    instance C Int where
+      f @Bool = ...
+
+  Rejected with the following message::
+
+    • Type indexes must match class instance head
+      Expected: f @Int
+        Actual: f @Bool
+    • In the method definition for ‘f’
+      In the instance declaration for ‘C Int’
+
+  This is the same message as one would get for associated types.
+  The implementation may opt to provide a different error message in the same spirit.
+
+Effect and Interactions
+-----------------------
```

Comment:
This section should mention whether there's any interaction with `-XInstanceSigs`, which one might reasonably expect to be affected